### PR TITLE
[Docker driver] - Adding support for sysctl and ulimit configuration

### PR DESCRIPTION
**Pull request by samber**
```hcl
task "db" {
    driver = "docker"
    config {
      image: "redis:3.2"
      sysctl {
        net.core.somaxconn = "16384"
      }
      ulimit {
        nproc = "4242"
        nofile = "2048:4096"
      }
    }
}
```

### CODE REVIEW

File: client/driver/docker.go [hunk 1]
```diff
@@ -142,6 +142,10 @@ type DockerDriverConfig struct {
 	PortMapRaw       []map[string]int    `mapstructure:"port_map"`           //
 	PortMap          map[string]int      `mapstructure:"-"`                  // A map of host port labels and the ports exposed on the container
 	Privileged       bool                `mapstructure:"privileged"`         // Flag to run the container in privileged mode
+	SysctlsRaw       []map[string]string `mapstructure:"sysctls"`            //
```
Comment (jippi) [inline client/driver/docker.go]: think the mapstructure here should be `sysctl` (singular) like all the other keys

File: client/driver/docker.go [hunk 2]
```diff
@@ -142,6 +142,10 @@ type DockerDriverConfig struct {
...
+	Sysctls          map[string]string   `mapstructure:"-"`                  // The sysctl custom configurations
```
Comment (jippi) [inline client/driver/docker.go]: singular here too(?)

File: client/driver/docker.go [hunk 3]
```diff
@@ -142,6 +142,10 @@ type DockerDriverConfig struct {
...
+	UlimitsRaw       []map[string]string `mapstructure:"ulimits"`            //
```
Comment (jippi) [inline client/driver/docker.go]: and here?

File: client/driver/docker.go [hunk 4]
```diff
@@ -142,6 +142,10 @@ type DockerDriverConfig struct {
...
+	Ulimits          []docker.ULimit     `mapstructure:"-"`                  // The ulimit custom configurations
```
Comment (jippi) [inline client/driver/docker.go]: and here?

**Comment by Ashald** [comment]
Is there is reason this must be docker specific? It'd be nice if was supported for all drivers (or at least exec/orther containers), xref: #2352

**Comment by samber** [comment]
@jippi made some fixes

CI does not pass, I think it's due to master branch

**Comment by samber** [comment]
rebased, CI passing

**Comment by lfarnell** [comment]
I would agree with @Ashald on this. If you're going to do it for container's at the least, you should do it for rkt as well. Not to mention other drivers would be benefit from this.

**Comment by samber** [comment]
@lfarnell @Ashald ok, some PRs should follow this one very soon, for other drivers

**Comment by dadgar** [comment]
Hey all, we are looking into adding a new stanza so this can be abstracted to all relevant drivers in 0.6.0. There won't be much activity on this PR until we have that design finalized.

**Comment by nanoz** [comment]
Any news on this @dadgar regarding the design of the sysctl stanza ? Not being able to configure somaxconn params and the like is a show stopper for production use

**Comment by samber** [comment]
If I can help in any way, just let me know ;)

**Comment by nanoz** [comment]
Hello @dadgar, any update on this issue? Do you now know what this generic stanza would look like?

We are really looking for this feature :)

**Comment by commarla** [comment]
Hey, 
Do you have any news on this ? Thanks.

**Comment by commarla** [comment]
Hey @dadgar any news on this ? Thanks

**Comment by jippi** [comment]
will `sysctl` and `ulimit` complain if you are trying to change a property that isn't within a cgroup? 

as an example redis expect `vm.overcommit_memory=1`, which isn't currently implemented as a cgroup property, and thus will affect the entire host system - i would not want those to be applied outside my containers scope, and thus those should fail hard rather than being silently ignored or even applied

**Comment by commarla** [comment]
Hi @jippi, 

no it doesn't, if I start my container in privileged mode and run `sysctl vm.overcommit_memory=1`. It affects the whole system and not only the container.

**Comment by jippi** [comment]
@commarla thats fine and expected, but what will the behaviour be if you start the container in non-privileged mode? just silently ignored?

**Comment by commarla** [comment]
@jippi it fails  with an error

```
root@439a24aaa498:/data# sysctl vm.overcommit_memory=1
sysctl: setting key "vm.overcommit_memory": Read-only file system
root@439a24aaa498:/data# echo $?
255
```

**Comment by jippi** [comment]
same behaviour if you do like below?

```
task "redis" {
    driver = "docker"
    config {
      image: "redis:3.2"
      sysctl {
        vm.overcommit_memory = "1"
      }
    }
}
```

**Comment by commarla** [comment]
I don't understand. `sysctl` is not a valid field. It is the purpose of this PR.

**Comment by jippi** [comment]
yes, my question is, what will the behaviour be once this PR is merged in, when providing `sysctl{}` that can't be applied within a cgroup, or that is launched without priviledged access

**Comment by commarla** [comment]
Ok the result should be the same than running this command 
```
❯ docker run --rm --sysctl="vm.overcommit_memory=1" redis:3.2
invalid argument "vm.overcommit_memory=1" for --sysctl=vm.overcommit_memory=1: sysctl 'vm.overcommit_memory=1' is not whitelisted
See 'docker run --help'.
```
docker only allow whitelisted sysctls options, see at the end of these page https://docs.docker.com/engine/reference/commandline/run/#configure-namespaced-kernel-parameters-sysctls-at-runtime

With a whitelisted sysctl it works:
```
docker run --rm --sysctl="net.core.somaxconn=16384" redis:3.2
Unable to find image 'redis:3.2' locally
3.2: Pulling from library/redis
d13d02fa248d: Pull complete
039f8341839e: Pull complete
```

**Comment by jippi** [comment]
great - couldn't be happier with this PR then :)

**Comment by nanoz** [comment]
How about merging this as is and create a second PR with rkt support later? 
Right now, I need to maintain a prestart script and run containers as privileged because of this missing feature and Im not confortable with the level of security it implies.

**Comment by Ashald** [comment]
We're using `rkt` and this feature is important for us as well so I hope we won't postpone `rkt` too much. :)

**Comment by preetapan** [comment]
Closing this in favor of #3568 where I rebased and fixed merge conflicts. That PR preserves commit history

**Pull request closed without merging**